Bug/reimplement headers

jphenow opened the pull request:
Addresses #162 

Open issue with axios test which is 500ing on header change. I don't know how to address the test but my notes may explain why the change would have broken something.

* Firefox's MDN docs spec for Headers is slightly different from `lib.dom.d`
  - Namely MDN specs `entries`, `keys`, and `values` functions
  - Weirdly, while the spec doesn't call out `forEach` it _is_ available on Chrome and Firefox
  - Chrome's forEach value param lists values as an array of string, while Firefox's is a string of comma-separated values
  - node's implementation matches Firefox more closely in terms of `forEach`
  - Furthermore, lib.dom (which I believe matches Node's actual implementation) doesn't specify the `entries`, `keys`, or `values` functions, but _does_ specify a `forEach`
* This implementation matches the `lib.dom` description and follows the Firefox/node spec of string, with comma separated values

Note: Values will only be comma separated automatically if using `headers.append`.

So, the open question is how I go about fixing the test that's broken. Also is there documentation, or anything that needs further update?

mrkurt commented:
That was one of our old school, hard to understand tests. It was failing because the `cookie_jar` code we use expects `headers.getAll`. I added a more useful test in v8env.

jphenow commented:
I got it running again with the reboot, but test output is a little more chatty with server output. Will have to dig in some more, but they look like tests related to Cookies as Kurt pointed out a few days ago, so not surprising failures at all. The header API is a slightly different from previously, so that's to be expected.

michaeldwan commented:
Yeah the test output is way too verbose now since the e2e tests are using the local server which prints to STDOUT. We'll be looking into that soon.

mrkurt commented:
@jphenow I think if the header just implements `getAll`, and maybe marks it deprecated like MDN, the cookie stuff will work fine.

mrkurt commented:
Also the test reporter in CI gives much nicer output than the console: https://dev.azure.com/flydotio/fly/_build/results?buildId=59&_a=summary&tab=ms.vss-test-web.test-result-details&view=ms.vss-test-web.test-result-details

jphenow commented:
Yea that's what I arrived at locally - just haven't had a chance to push and see what happens. I figure the lib.dom.d can present an API that's generally available and you can manually import the type if you want the added API bits. At least this way it'll be a superset rather than wholly different from lib.dom.d which is nice. 

I did run into some other strange issues, so may pair at some point and see if we can improve some more of the internals here, but we'll see what happens after I do what I said I would in paragraph 1 😄

michaeldwan commented on packages/v8env/src/cache.ts:
```diff
@@ -56,13 +56,9 @@ export default {
-        resHeaders[h[0]] = (h[1].join && h[1].join(",")) || h[1]
-      }
-    }
+    res.headers.forEach((value, name) => {
```
Prefer `for ... of` over `forEach` since we're targeting ES6+

```javascript
for (const [name, value] of res.headers) {
 ...
}
```

michaeldwan commented on packages/v8env/src/headers.ts:
```diff
@@ -10,175 +10,116 @@ import { logger } from "./logger"
-  if (init) {
-    fill(this, init)
+export default class Headers {
+  private headerMap: { [key: string]: string[] } = {};
```
We can use `Map<string, string[]>` which gives native `has`, `forEach`, and iterator.

michaeldwan commented on packages/v8env/src/headers.ts:
```diff
@@ -10,175 +10,116 @@ import { logger } from "./logger"
+    return { ...this.headerMap };
   }
-}
-Headers.prototype[Symbol.iterator] = function() {
```
We want to keep the iterator for `for ... of` support

michaeldwan commented on packages/v8env/src/headers.ts:
```diff
@@ -10,175 +10,116 @@ import { logger } from "./logger"
-        init[name].forEach(v => {
-          headers.append(name, v)
-        })
+  public forEach(callbackfn: (value: string, key: string, parent: Headers) => void, thisArg?: any): void {
```
Nice addition

jeromegn commented on packages/v8env/src/headers.ts:
```diff
@@ -10,175 +10,116 @@ import { logger } from "./logger"
-  })
-  if (init) {
-    fill(this, init)
+export default class Headers {
```
This could benefit from stronger typing by using `implements` like:

```typescript
export default class FlyHeaders implements Headers {
```

It can still implement more functions, but at least we'd know we support the spec. (assuming typescript's dom types are accurate)

jphenow commented:
Added back `for..of` Iterator support, updated external use of iterator to `for..of` and now it implements the Header interface.

Implementing Header interface meant that the class needed to be renamed (since I can't rename the global within scope, I don't believe), which also meant I had to assert type to `FlyHeaders` in order to get `for..of` since Iterator isn't specified in the `lib.dom` interface. Sort of awkward without controlling that interface or at least taking some control of it.

michaeldwan commented:
Okay I see now, [`Headers` in `lib.dom`](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.dom.d.ts#L8546) doesn't support iteration, you need to include [`lib.dom.iterable`](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.dom.iterable.d.ts#L103) for that. We'll need to add that to v8env's tsconfig then you don't need to cast. That said, I think it's better for v8env code to explicitly import modules instead of relying on the global declarations, so your change to `FlyHeaders` is great.

michaeldwan commented on packages/v8env/src/fly/cache/response.ts:
```diff
@@ -129,11 +130,10 @@ export async function set(key: string, resp: Response, options?: ResponseCacheSe
-  for (const headerSet of resp.headers as any) {
-    const [name, value] = headerSet
+  const skipHeaderSet = new Set(skipHeaderOption);
+  for (const [value, name] of resp.headers as FlyHeaders) {
```
Can remove the cast

michaeldwan commented on v8env/test/cookie_jar.spec.js:
```diff
@@ -0,0 +1,23 @@
+import { expect } from 'chai'
```
This file is in the old test path, needs to be in `tests/v8env/tests/` to be ran

jphenow commented:
Alrighty. Pushed some updates, sorry for slowness getting this one wrapped up. Pretty sure there's still some wonkiness going on. Will take another look this evening hopefully

jphenow commented:
Well, those are weird

mrkurt commented:
Wow that is really strange.

jphenow commented:
![](https://media.giphy.com/media/l3V0sNZ0NGomeurCM/giphy.gif)

jphenow commented:
We cool to close this then? At long last?

michaeldwan commented:
🎉 thanks @jphenow!

jphenow commented:
Wow was I sleeping at the wheel? Thanks for reading through the changes 🤦‍♂️

michaeldwan commented:
It's all good 😄
